helper/schema/resource_test.go

```diff
@@ -1368,3 +1373,111 @@ func TestResourceData_timeouts(t *testing.T) {
 		t.Fatalf("incorrect ResourceData timeouts: %#v\n", *data.timeouts)
 	}
 }
+
+func TestResource_UpgradeState(t *testing.T) {
+	// Schema v2 it deals only in newfoo, which tracks foo as an int
+	r := &Resource{
+		SchemaVersion: 2,
+		Schema: map[string]*Schema{
+			"newfoo": &Schema{
+				Type:     TypeInt,
+				Optional: true,
+			},
+		},
+	}
+
+	r.LegacySchema.Version = 1
+	r.LegacySchema.Type = cty.Object(map[string]cty.Type{
+		"id":     cty.String,
+		"oldfoo": cty.Number,
+	})
+
+	r.UpgradeState = func(
+		v int,
+		m map[string]interface{},
+		meta interface{}) (map[string]interface{}, error) {
+
+		oldfoo, ok := m["oldfoo"].(float64)
+		if !ok {
+			t.Fatalf("expected 1.2, got %#v", m["oldfoo"])
+		}
+		m["newfoo"] = int(oldfoo * 10)
+		delete(m, "oldfoo")
+
+		return m, nil
+	}
+
+	oldStateAttrs := map[string]string{
+		"id":     "bar",
+		"oldfoo": "1.2",
+	}
+
+	// convert the legacy flatmap state to the json equivalent
+	val, err := hcl2shim.HCL2ValueFromFlatmap(oldStateAttrs, r.LegacySchema.Type)
+	if err != nil {
+		t.Fatal(err)
+	}
+	js, err := ctyjson.Marshal(val, r.LegacySchema.Type)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// unmarshal the state using the json default types
+	var m map[string]interface{}
+	if err := json.Unmarshal(js, &m); err != nil {
+		t.Fatal(err)
+	}
+
+	actual, err := r.UpgradeState(2, m, nil)
```

Comment:
Yeah, this is almost more of an example than anything. Because this is going to be called directly from core, there really isn't any `helper/schema` code that will interact with it. It was also an exercise in translating the existing `MigrateState` test to get the same behavior.

Reply:
I assume in the end `helper/schema` will have the implementation of the `UpgradeResourceState` method which will be the entry point for core's call, and then that'd be a "front door" to call from this test, but indeed that's for later. This LGTM to illustrate how this is meant to work.